### PR TITLE
mgr/mgr_util: switch using unshared cephfs connections whenever possible

### PR DESCRIPTION
**Pull request by vshankar**
The connection pool used by mgr/volumes (mgr/snap_schedule, mgr/mirroring)
is basically a cached libcephfs connection handle. Since its a singleton
lock contention may occur in the client (contention on `client_lock`),
especially when an MDS operation taken too long to complete.

Having a pool of connections to a cephfs file system, ensures that concurrent
access to ceph file system via the connection pool use unshared libcephfs
handles whenever possible thereby not contending for `client_lock`. However,
there is a max cap of unshared connections per cephfs file system, post which
connections start to get shared between users.

Note that to totally benefit from this, its desirable to have multiple active
MDSs for a ceph file system to rule out the possibility of the contention in
`mds_lock`.

Fixes: http://tracker.ceph.com/issues/51256
Signed-off-by: Venky Shankar <vshankar@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/mgr_util.py [hunk 1]
```diff
@@ -194,57 +195,86 @@ def __init__(self, mgr: Module_T):
...
-                conn = self.connections.get(fs_name, None)
-                if conn:
-                    if conn.is_connection_valid():
-                        return conn.get_fs_handle()
+                min_shared = 0
+                shared_connection = None
+                connections = self.connections.get(fs_name, [])
+                logger.debug(f'[get] volume: ({fs_name}) connection: ({connections})')
+                if not connections:
+                    self.connections[fs_name] = []
```
Comment (batrick) [inline src/pybind/mgr/mgr_util.py:216]:
```suggestion
```

**Comment by vshankar** [comment]
> As we discussed on today's call: let's open a ticket about maybe testing a dedicated libcephfs handle for subvolume API calls. That would potentially improve cache efficiency in the volumes plugin.

Makes sense. ACK.

File: src/pybind/mgr/mgr_util.py [hunk 2]
```diff
@@ -194,57 +195,84 @@ def __init__(self, mgr: Module_T):
...
+            for connection in connections:
+                if connection.fs == fs_handle:
+                    logger.debug(f'[put] connection: {connection} usage: {connection.ops_in_progress}')
+                    connection.put_fs_handle(notify=lambda: self.cond.notifyAll())
 
-    def _del_fs_handle(self, fs_name: str, wait: bool = False) -> None:
-        conn = self.connections.pop(fs_name, None)
-        if conn:
-            conn.del_fs_handle(waiter=None if not wait else lambda: self.cond.wait())
+    def _del_connection(self, fs_name: str, connection: Connection, wait: bool = False) -> None:
```
Comment (ajarr) [inline src/pybind/mgr/mgr_util.py:259]: The method that this replaces, _del_fs_handle, handles the case where 'fs_name' key is not present (`conn = self.connections.pop(fs_name, None)`) in self.connections, and deletes fs handle only if connection is non-empty (`if conn:`). Both these cases need not be handled anymore?

Comment (vshankar) [reply src/pybind/mgr/mgr_util.py:259]: it's kind of handled by returning an empty list if the key is not present which turn the loop iteration to a no-op.

**Review by ajarr — APPROVED** [review]

File: src/pybind/mgr/mgr_util.py [hunk 3]
```diff
@@ -194,57 +195,84 @@ def __init__(self, mgr: Module_T):
...
+                            self._del_connection(fs_name, connection)
                     else:
-                        # filesystem id changed beneath us (or the filesystem does not exist).
-                        # this is possible if the filesystem got removed (and recreated with
-                        # same name) via "ceph fs rm/new" mon command.
-                        logger.warning("filesystem id changed for volume '{0}', reconnecting...".format(fs_name))
-                        self._del_fs_handle(fs_name)
-                conn = CephfsConnectionPool.Connection(self.mgr, fs_name)
-                conn.connect()
+                        if min_shared < connection.ops_in_progress:
```
Comment (kotreshhr) [inline src/pybind/mgr/mgr_util.py:229]: Doesn't this return max shared connection ?

Comment (vshankar) [reply src/pybind/mgr/mgr_util.py:229]: it does -- needs flip.

**Comment by vshankar** [comment]
@kotreshhr fixed and updated. please check.

File: src/pybind/mgr/mgr_util.py [hunk 4]
```diff
@@ -194,57 +195,86 @@ def __init__(self, mgr: Module_T):
...
-                conn = self.connections.get(fs_name, None)
-                if conn:
-                    if conn.is_connection_valid():
-                        return conn.get_fs_handle()
+                min_shared = 0
+                shared_connection = None
+                connections = self.connections.setdefault(fs_name, [])
+                logger.debug(f'[get] volume: ({fs_name}) connection: ({connections})')
+                if connections:
+                    min_shared = connections[0].ops_is_progress
```
Comment (kotreshhr) [inline src/pybind/mgr/mgr_util.py:216]: shared_connection should also be initialized here to connections[0].

Comment (vshankar) [reply src/pybind/mgr/mgr_util.py:216]: I think I did that :/

**Review by kotreshhr — APPROVED** [review]

**Review by batrick — CHANGES_REQUESTED** [review]
````
mgr_util.py: note: In member "get_fs_handle" of class "CephfsConnectionPool":
mgr_util.py:216: error: "Connection" has no attribute "ops_is_progress"; maybe "ops_in_progress"?
Found 1 error in 1 file (checked 29 source files)
````

**Comment by vshankar** [comment]
> ```
> mgr_util.py: note: In member "get_fs_handle" of class "CephfsConnectionPool":
> mgr_util.py:216: error: "Connection" has no attribute "ops_is_progress"; maybe "ops_in_progress"?
> Found 1 error in 1 file (checked 29 source files)
> ```

:/ fixed

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-June-26

**Pull request merged**